orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/cloudformation/EvaluateCloudFormationChangeSetExecutionTask.java

```diff
@@ -0,0 +1,116 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.tasks.providers.aws.cloudformation;
+
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.OverridableTimeoutRetryableTask;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+public class EvaluateCloudFormationChangeSetExecutionTask
+    implements OverridableTimeoutRetryableTask {
+
+  @Override
+  public @Nonnull TaskResult execute(Stage stage) {
+    String actionOnReplacement =
+        (String) Optional.ofNullable(stage.getContext().get("actionOnReplacement")).orElse("ask");
+
+    if (!actionOnReplacement.equals("ask")) {
+      return TaskResult.SUCCEEDED;
+    }
+
+    Optional<Map> currentChangeSet = getCurrentChangeSet(stage);
+    Optional<Boolean> changeSetIsReplacement =
+        Optional.ofNullable(getChangeSetIsReplacement(stage));
+
+    if (!changeSetIsReplacement.isPresent()) {
+      Boolean isReplacement;
+      if (isAnyChangeSetReplacement(currentChangeSet.get())) {
+        isReplacement = true;
+      } else {
+        isReplacement = false;
+      }
+      Map<String, Object> context = stage.getContext();
+      context.put("changeSetIsReplacement", isReplacement);
+      return TaskResult.builder(ExecutionStatus.RUNNING).context(context).build();
+    }
+
+    if (!changeSetIsReplacement.orElse(false)) {
+      return TaskResult.SUCCEEDED;
+    }
+
+    String changeSetExecutionChoice = getChangeSetExecutionChoice(stage);
+    if (changeSetExecutionChoice != "none") {
+      Map<String, Object> context = stage.getContext();
+      context.put("actionOnReplacement", changeSetExecutionChoice);
+      return TaskResult.builder(ExecutionStatus.SUCCEEDED).context(context).build();
+    }
+    return TaskResult.RUNNING;
+  }
+
+  @Override
+  public long getBackoffPeriod() {
+    return 1_000;
+  }
+
+  @Override
+  public long getTimeout() {
+    return Integer.MAX_VALUE;
+  }
+
+  private String getChangeSetExecutionChoice(Stage stage) {
+    return (String)
+        Optional.ofNullable(stage.getContext().get("changeSetExecutionChoice")).orElse("none");
+  }
+
+  private Boolean getChangeSetIsReplacement(Stage stage) {
```

Comment:
You are returning a null here if the key is not there, and then transforming it back to optional on the caller. Maybe return Optionals on all these helper methods and react accordingly on the main function? We could get rid of the "none" string with is just signaling a non-existing value.

Reply:
I've changed the logic of this. WDYT?